### ISSUE TITLE
Address feedback from conversations with `asyncio-mqtt` author

### ISSUE THREAD

**Issue by bachya** [bachya/ecowitt2mqtt#233]
[A discussion with the author of `asyncio-mqtt`](https://github.com/sbtinstruments/asyncio-mqtt/issues/120) revealed some low-level bugs that need fixing:

- [x] We need to create a new `Client` object each time we reconnect #235
- [x] We should keep the connection open (and create our own retry logic) vs. opening a new connection with each publish https://github.com/bachya/ecowitt2mqtt/pull/236
~- [ ] We should be smarter about how server payload callbacks are implemented~

**Comment by bachya** [comment]
Not going to do the second one right now. The nature of Uvicorn + FastAPI (and the whole respond-to-payload-events architecture) makes keeping a single client prohibitive. As long as we'll properly canceling tasks, I'm not overly concerned about a fresh connect/disconnect with each publish.

**Comment by Anto79-ops** [comment]
Thanks Aaron, I don't really mind the elevated CPU usage, which I've noticed that ecowitt2mqtt does do, even when working properly for me, but I think if we solve problem 1 that'll resolve the issue of the choked CPU on my RPi. That's the issue that I'm after here.

**Comment by bachya** [comment]
(3) will actually be fixed by (1), so we have less to do that I thought.

**Comment by Anto79-ops** [comment]
thanks for being so quick to address issues. I just updated to 2022.7.6 and restarted my RPi. Fingers crossed :)

**Comment by alexdelprete** [comment]
> As long as we'll properly canceling tasks, I'm not overly concerned about a fresh connect/disconnect with each publish.

I had problems with the weewx-mqtt driver in the past, it would create problems with some brokers, because it connected/disconnected at every message. luckily someone made a PR for the driver to keep up the connection and that solved all the issues.

Mosquitto devs (and other MQTT brokers advise the same thing) [suggest](https://twitter.com/MosquittoMQTT/status/1293935424818352130) to connect only once, not once per message, particularly with encrypted connections to the broker:

![image](https://user-images.githubusercontent.com/7027842/181655299-e750d0fe-2ed0-484e-9a28-9e6319cc3efc.png)

But from what I understand, the architecture of ecowitt2mqtt doesn't allow you to manage only one connection. I hope you'll find a solution for the future.

**Comment by bachya** [comment]
I found a solution. Will work on it tomorrow. 👍🏻

**Comment by alexdelprete** [comment]
I knew you would...:)

**Comment by alexdelprete** [comment]
@bachya Aaron, reading this on uvicorn repo leads me to think it will never be a perfect fit for single connection, they write "less of a limiting factor" in this description:

![image](https://user-images.githubusercontent.com/7027842/181801289-92b9a9af-480e-43a6-b593-296749778768.png)

Brainstorming a little bit here, correct me if I'm wrong: uvicorn is ASGI, but deriving from WSGI, it has always a "transactional" approach. Having a permanent connection with a broker is a little out-of-scope respect to what that kind of gateway was intended for. Unless there are features that allow you to have permanent "workers" (or threads) to which you can delegate the connection to the MQTT broker and refer to that worker in the main app.

**Comment by bachya** [comment]
@alexdelprete The difference lies in the fact that we are using `asyncio` rather than a threaded worker model. When I speak of two tasks – an MQTT process loop and the Uvicorn REST API server – those are two tasks in a cooperative multitasking framework, not two workers from the more traditional threading world.

In that sense, Uvicorn should handle multiple `asyncio` tasks just fine. The problem is that they are catching the signals that normally terminate their server (Docker image stopping, Ctrl-C, etc.) and swallowing them; they should be doing their shutdown and then raising the traditional `asyncio.CancelledError` exception for downstream applications (like us) to (a) learn that the server has shut down and (b) perform their own cleanup tasks. https://github.com/encode/uvicorn/issues/1579 speaks to this: the author is asking Uvicorn not to swallow those signals.

Does that make sense?

**Comment by alexdelprete** [comment]
Yes it does, thanks.

It seems like a lack of implementation of something quite important, I'm wondering: maybe it lacks that because they didn't feel it would've been used in a specific way.

Questions:

1. can you create/use a threaded worker inside an asyncio task?
2. can you have a worker thread that starts an asyncio task?

Just trying to understand the flexibility of uvicorn implementation.

**Comment by bachya** [comment]
Yep, all good questions.

> 1. can you create/use a threaded worker inside an asyncio task?

Yes: you simply pass the function containing the threaded work to [`loop.run_in_executor`](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor).

> 2. can you have a worker thread that starts an asyncio task?

Within a non-async function, you can use `asyncio.run()` to create a task. However, care must be taken to know which event loop is accessed by which thread.

> 3. you avoided the threaded worker model because it was more complex to manage for the scope?

In essence, yes. Mixing sync and async models:

* adds a lot of overhead to manage.
* creates potentially tricky bugs if a thread attempts to access the wrong event loop (and vice versa).
* eliminates the speed, consistency, and stability of a model that uses one or the other (I'm partial to async, so all of the libraries used here are async-focused).

**Comment by alexdelprete** [comment]
> Within a non-async function, you can use `asyncio.run()` to create a task. However, care must be taken to know which event loop is accessed by which thread.

So you could use that to create the permanent session to the MQTT broker and keep the task under control via the thread management specific events/functions. Using a thread would bypass the limitation of the missing signals, right?

**Comment by Anto79-ops** [comment]
it might be too early to say for certain, but with the new code, I have NOT experienced any hiccups, my RPi is just humming away. Its been 19 hrs, the real test would be to give it 5 days. Here's the last 24 hrs, that spike in CPU was doing updates and when I installed the new code, so after that spike, is running the 2022.7.6.

![image](https://user-images.githubusercontent.com/79593761/181812302-0d8acb01-422b-4cc8-a862-42b78570fff0.png)

if @alexdelprete and @bachya you are discussing ways to make the integration less CPU intensive by the above, that would be good too. It seem when ecowitt2mqtt runs on my RPi properly, it can tax the CPU by 6 to 12% extra (and this when it run properly, and not choke the CPU).

**Comment by alexdelprete** [comment]
> you are discussing ways to make the integration less CPU intensive by the above, that would be good too.

we're discussing about how to improve efficiency: in an MQTT scenario, the very frequent connect/disconnect for every message is not efficient, and Aaron regonised that. Now the issue is how to best implement the permanent connection to the MQTT broker, given the architecture of Ecowitt2MQTT, Aaron is a good developer, he will find the best solution, given the constraints.

I'm confident that 12% is too high even for a raspberry, it might be acceptable but I'm very curious to see when Aaron implements the permanent connection the results, we'll use your RPi as a benchmark. :)

**Comment by Anto79-ops** [comment]
@alexdelprete You're right, Aaron always figures stuff out! 

By TLS, do you mean username/password? No, my mosquito MQTT broker does not have any security to it. Any device can connect to it without username and/or password field on broker port.

**Comment by Anto79-ops** [comment]
How can I check? I'm running mosquitto version 2.0.11 on my Ubuntu 22.04 Server.

**Comment by bachya** [comment]
Thank you both for the kind words – it makes this work very enjoyable!

@alexdelprete

> So you could use that to create the permanent session to the MQTT broker and keep the task under control via the thread management specific events/functions. Using a thread would bypass the limitation of the missing signals, right?

That's right. It would take more overhead (I'd have to manage a separate set of signal handlers for the worker thread), but it would be feasible. That said, I've figured out a way to patch Uvicorn's `Server` object to do what I need in a fully async fashion. Stay tuned.

@Anto79-ops

> How can I check? I'm running mosquitto version 2.0.11 on my Ubuntu 22.04 Server.

To use TLS, you would (a) run your `ecowitt2mqtt` instance with `--mqtt-tls` and (b) have configured Mosquitto to validate and accept TLS connections.

Sidenote, can you remind me of the specs of your Pi (generation, RAM, etc.)?

**Comment by Anto79-ops** [comment]
Thanks for clearing that up!

I definitely do not use --mqtt-tls and have not configured Mosquitto to validate and accept TLS connections.

As for my RPi, here's the information:

![image](https://user-images.githubusercontent.com/79593761/181832027-22d4d513-76c1-4e19-9a74-c2bb6a00b040.png)

**Comment by bachya** [comment]
#236 is looking good. Before I go down the road of adding tests/etc., I'd love some testing. @Anto79-ops Since you install straight from PyPI, try installing that version:

```
pip install git+https://github.com/bachya/ecowitt2mqtt.git@single-mqtt-connection
```

...and let us know how it goes?

**Comment by Anto79-ops** [comment]
I would be happy too!

Here's what I did and see:

```
pi@raspberrypi:~ $ sudo service ecowitt2mqtt stop
pi@raspberrypi:~ $ sudo pip install git+https://github.com/bachya/ecowitt2mqtt.g                                                                                                                                                                                                                                                                                                                                                                                                             it@single-mqtt-connection
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting git+https://github.com/bachya/ecowitt2mqtt.git@single-mqtt-connection
  Cloning https://github.com/bachya/ecowitt2mqtt.git (to revision single-mqtt-connection) to /tmp/pip-req-build-dmny5q3q
  Running command git clone -q https://github.com/bachya/ecowitt2mqtt.git /tmp/pip-req-build-dmny5q3q
  Running command git checkout -b single-mqtt-connection --track origin/single-mqtt-connection
  Switched to a new branch 'single-mqtt-connection'
  Branch 'single-mqtt-connection' set up to track remote branch 'single-mqtt-connection' from 'origin'.
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Requirement already satisfied: uvloop<0.17.0,>=0.16.0 in /usr/local/lib/python3.9/dist-packages (from ecowitt2mqtt==2022.7.6) (0.16.0)
Requirement already satisfied: fastapi<0.79.0,>=0.78.0 in /usr/local/lib/python3.9/dist-packages (from ecowitt2mqtt==2022.7.6) (0.78.0)
Requirement already satisfied: typer[all]<0.5.0,>=0.4.1 in /usr/local/lib/python3.9/dist-packages (from ecowitt2mqtt==2022.7.6) (0.4.1)
Requirement already satisfied: uvicorn<0.19.0,>=0.18.0 in /usr/local/lib/python3.9/dist-packages (from ecowitt2mqtt==2022.7.6) (0.18.2)
Requirement already satisfied: ruamel.yaml<0.18.0,>=0.17.21 in /usr/local/lib/python3.9/dist-packages (from ecowitt2mqtt==2022.7.6) (0.17.21)
Requirement already satisfied: thefuzz[speedup]<0.20.0,>=0.19.0 in /usr/local/lib/python3.9/dist-packages (from ecowitt2mqtt==2022.7.6) (0.19.0)
Requirement already satisfied: meteocalc<2.0.0,>=1.1.0 in /usr/local/lib/python3.9/dist-packages (from ecowitt2mqtt==2022.7.6) (1.1.0)
Requirement already satisfied: asyncio-mqtt>=0.12.1 in /usr/local/lib/python3.9/dist-packages (from ecowitt2mqtt==2022.7.6) (0.12.1)
Requirement already satisfied: python-multipart<0.0.6,>=0.0.5 in /usr/local/lib/python3.9/dist-packages (from ecowitt2mqtt==2022.7.6) (0.0.5)
Requirement already satisfied: paho-mqtt>=1.6.0 in /usr/local/lib/python3.9/dist-packages (from asyncio-mqtt>=0.12.1->ecowitt2mqtt==2022.7.6) (1.6.1)
Requirement already satisfied: starlette==0.19.1 in /usr/local/lib/python3.9/dist-packages (from fastapi<0.79.0,>=0.78.0->ecowitt2mqtt==2022.7.6) (0.19.1)
Requirement already satisfied: pydantic!=1.7,!=1.7.1,!=1.7.2,!=1.7.3,!=1.8,!=1.8.1,<2.0.0,>=1.6.2 in /usr/local/lib/python3.9/dist-packages (from fastapi<0.79.0,>=0.78.0->ecowitt2mqtt==2022.7.6) (1.9.1)
Requirement already satisfied: anyio<5,>=3.4.0 in /usr/local/lib/python3.9/dist-packages (from starlette==0.19.1->fastapi<0.79.0,>=0.78.0->ecowitt2mqtt==2022.7.6) (3.6.1)
Requirement already satisfied: typing-extensions>=3.10.0 in /usr/local/lib/python3.9/dist-packages (from starlette==0.19.1->fastapi<0.79.0,>=0.78.0->ecowitt2mqtt==2022.7.6) (4.2.0)
Requirement already satisfied: sniffio>=1.1 in /usr/local/lib/python3.9/dist-packages (from anyio<5,>=3.4.0->starlette==0.19.1->fastapi<0.79.0,>=0.78.0->ecowitt2mqtt==2022.7.6) (1.2.0)
Requirement already satisfied: idna>=2.8 in /usr/lib/python3/dist-packages (from anyio<5,>=3.4.0->starlette==0.19.1->fastapi<0.79.0,>=0.78.0->ecowitt2mqtt==2022.7.6) (2.10)
Requirement already satisfied: six>=1.4.0 in /usr/lib/python3/dist-packages (from python-multipart<0.0.6,>=0.0.5->ecowitt2mqtt==2022.7.6) (1.16.0)
Requirement already satisfied: ruamel.yaml.clib>=0.2.6 in /usr/local/lib/python3.9/dist-packages (from ruamel.yaml<0.18.0,>=0.17.21->ecowitt2mqtt==2022.7.6) (0.2.6)
Requirement already satisfied: python-levenshtein>=0.12 in /usr/local/lib/python3.9/dist-packages (from thefuzz[speedup]<0.20.0,>=0.19.0->ecowitt2mqtt==2022.7.6) (0.12.2)
Requirement already satisfied: setuptools in /usr/lib/python3/dist-packages (from python-levenshtein>=0.12->thefuzz[speedup]<0.20.0,>=0.19.0->ecowitt2mqtt==2022.7.6) (52.0.0)
Requirement already satisfied: click<9.0.0,>=7.1.1 in /usr/local/lib/python3.9/dist-packages (from typer[all]<0.5.0,>=0.4.1->ecowitt2mqtt==2022.7.6) (8.1.3)
Requirement already satisfied: colorama<0.5.0,>=0.4.3 in /usr/local/lib/python3.9/dist-packages (from typer[all]<0.5.0,>=0.4.1->ecowitt2mqtt==2022.7.6) (0.4.5)
Requirement already satisfied: shellingham<2.0.0,>=1.3.0 in /usr/local/lib/python3.9/dist-packages (from typer[all]<0.5.0,>=0.4.1->ecowitt2mqtt==2022.7.6) (1.4.0)
Requirement already satisfied: h11>=0.8 in /usr/local/lib/python3.9/dist-packages (from uvicorn<0.19.0,>=0.18.0->ecowitt2mqtt==2022.7.6) (0.13.0)
pi@raspberrypi:~ $
```
Do I simply just start the service again to run the new updates?

**Comment by bachya** [comment]
@Anto79-ops Correct!

**Comment by Anto79-ops** [comment]
@bachya Awesome!

Its running as we speak....keep you posted :) 

Cheers

**Comment by Anto79-ops** [comment]
FYI. no error messages on startup.

![image](https://user-images.githubusercontent.com/79593761/181838939-fd44fd95-65cc-4bfe-a351-2880dc37d0b4.png)

**Comment by alexdelprete** [comment]
> #236 is looking good. Before I go down the road of adding unit tests/etc., I'd love some field testing. @Anto79-ops Since you install straight from PyPI, try installing that version:

Can I test it on docker? What tag?

**Comment by alexdelprete** [comment]
> That said, I've figured out a way to patch Uvicorn's `Server` object to do what I need in a fully async fashion. Stay tuned.

will you submit a PR to uvicorn? :)

**Comment by alexdelprete** [comment]
> Its running as we speak....keep you posted :)

let's also check that cpu graph...if you can change the Y-axis to 0-25 we could better check the small changes to the CPU usage. :)

**Comment by bachya** [comment]
> > #236 is looking good. Before I go down the road of adding unit tests/etc., I'd love some field testing. @Anto79-ops Since you install straight from PyPI, try installing that version:
> 
> Can I test it on docker? What tag?

Not yet – it will go into `dev` (and a tagged version) once merged. You can build the image yourself.

> > That said, I've figured out a way to patch Uvicorn's Server object to do what I need in a fully async fashion. Stay tuned.

> will you submit a PR to uvicorn? :)

We'll see... 😉

**Comment by alexdelprete** [comment]
> Not yet – it will go into `dev` (and a tagged version) once merged. You can build the image yourself.

I'll wait...:)

BTW, did you see [this](https://doc.ecowitt.net/web/#/apiv3en?page_id=1)? ;)

**Comment by bachya** [comment]
> > Not yet – it will go into `dev` (and a tagged version) once merged. You can build the image yourself.
> 
> I'll wait...:)
> 
> BTW, unrelated to the discussion, did you see [this](https://doc.ecowitt.net/web/#/apiv3en?page_id=1)? ;)

Interesting! Looks like it's a cloud-based API?

**Comment by alexdelprete** [comment]
> Interesting! Looks like it's a cloud-based API?

Yes...basically they finally created an API for the ecowitt.net service. There are many interesting things.

Another nice thing is the UDP API, GW1000 and GW2000 broadcast sensor data via UDP, port 45000. My RainMachine uses the GW1000 driver to get rain data and it uses UDP. The Weewx driver uses UDP too.

**Comment by Anto79-ops** [comment]
so far so good using the single mqtt connection code. Here's a CPU trace for the last 5 days or so.

![image](https://user-images.githubusercontent.com/79593761/181920337-4da739e1-5e0e-4bed-aa8f-f610de054a77.png)

 In places where the CPU spikes (chokes to 30%, or 1 core is 100%) and stops (i.e. pre 2022.07.6), that is me killing the service, otherwise it would just go on like that with elevated CPU. If we can go 5 days without a spiked CPU, I would consider this solved.

The RPi has other services running, so its hard to tell if the code is more or less CPU intensive....as long as it doesn't choke...im ok with that!

@bachya looking good so far!!

**Comment by Anto79-ops** [comment]
more good news, nothing new to report after 1 day and 21 hrs. since the update to single mqtt connection. CPU usage is hovering around 8% (average) and data is being reported from both gateways flawlessly!

**Comment by bachya** [comment]
That's great to hear, @Anto79-ops! Given this, I'm going to merge my PR and will likely publish one more July release today. Thanks for your testing!
